### ISSUE TITLE
S3 bucket name is unexpectedly being cut

### ISSUE THREAD

**Issue by colin-nolan** [zertrin/duplicity-backup.sh#155]
I'm using this script to help backup data to an internal "S3 compatible" object store (i.e. not AWS). 

The problem I hit is on these lines https://github.com/zertrin/duplicity-backup.sh/blob/418ab79b1131718030dd28518849c3c406935789/duplicity-backup.sh#L642-L648, where for some reason that I don't understand, the destination that is used with `s3cmd du [s3://BUCKET[/PREFIX]]`, gets cut such that the first 10 characters are removed?!

The destination I'm using is:
```
DEST="s3://consul-backup"
```

This debug output shows the destination that the script is trying to use:
```
---------[ Destination Disk Use Information ]--------
...
++ echo s3://consul-backup
++ cut -c 11-
+ TMPDEST=l-backup
...
Invoked as: /usr/bin/s3cmd -c /etc/opt/duplicity/.s3cfg du -H s3://l-backup
...
<Error follows as that bucket doesn't exist>
...
```

Manually invoking `/usr/bin/s3cmd -c /etc/opt/duplicity/.s3cfg du -H s3://consul-backup` (i.e. not chopping the first couple of characters from the bucket) works fine.

I have [a configuration for s3cmd](https://github.com/wtsi-hgi/hgi-systems/blob/6c0ccdcf572d3ce4dcfeb9eb6fab58816b3e19cd/ansible/roles/s3-backup/templates/s3cmd.ini.j2) such that it can be used (I assume) in the same was as if it was interacting with AWS. Therefore, I am confused as to why this doesn't work for me if others are using it without issues.

In case their useful:
- Template of my .s3cmd configuration file: https://github.com/wtsi-hgi/hgi-systems/blob/6c0ccdcf572d3ce4dcfeb9eb6fab58816b3e19cd/ansible/roles/s3-backup/templates/s3cmd.ini.j2
- Template of my duplicity-backup.sh.conf: https://github.com/wtsi-hgi/hgi-systems/blob/6c0ccdcf572d3ce4dcfeb9eb6fab58816b3e19cd/ansible/roles/s3-backup/templates/duplicity-backup-sh.conf.j2

**Comment by zertrin** [comment]
I think it is because of the default `s3+http://` scheme, which is 10 characters long. It looks like the other possible scheme `s3://` has been overlooked while writing this part of the code.

This commit removes an important part of the commentary in the conf file: https://github.com/zertrin/duplicity-backup.sh/commit/7bb45ff939c6bca3776950731a61d07f284f0d97#diff-cd68fa46e0aefe4cc5826f3ac7729e30L93

Previous comment that disappeared 4 years ago:

> NOTE: You do need to keep the "s3+http://<your location>/" format
> even though duplicity supports "s3://<your location>/".

The solution would be to improve this section of code to handle both cases and cut only up to the `://`.

**Comment by colin-nolan** [comment]
Thanks @zertrin for the information.

I changed the script such that it cuts the scheme correctly. However, fixing that reveals an issue with using the `s3://<bucket>` form with Duplicity:
```
$ /usr/bin/duplicity  -v3 --full-if-older-than 14D --s3-use-new-style  --no-encryption    /data/consul/data s3://consul-backup
BackendException: Boto requires a bucket name.
```

Taking a look at Duplicity's help shows that it reads the URL differently depending on the scheme: 
```
$ duplicity --help 
...

Backends and their URL formats:
...
  s3://other.host/bucket_name[/prefix]
  s3+http://bucket_name[/prefix]
...
```

I've not looked into why Duplicity decides on that interpretation and whether this is part of some kind of standard(?). 

If there is no real reason for this difference in s3 vs s3+http interpretation, duplicity-backup.sh should either reject destinations in the form s3://bucket_name (as it doesn't currently work with them) or convert them into a s3+http://bucket_name form.

**Comment by zertrin** [comment]
Hi,

> If there is no real reason for this difference in s3 vs s3+http interpretation

you might want to read the comments in the config file, the difference is already explained there:

> The s3+http scheme uses the default aws s3 hostname.
> Use s3://host/bucket/[backup-folder/] if you want to specify the host name.

So with `s3+http://` you must use AWS, and only with `s3://` you can choose another S3-compatible host.

Why this choice, I have no idea, but there is a reason between the 2 url schemes.

**Comment by zertrin** [comment]
Also note that:
 * the syntax `s3://bucket_name` ist not referenced anywhere, I don't know where you got it
 * the syntax `s3://host/bucket_name/` seems to be the correct one.

From Duplicity manpage:

>  S3 storage (Amazon)
> 
>     s3://host[:port]/bucket_name[/prefix]
>     s3+http://bucket_name[/prefix]

Note the `host` part is only present in the `s3://` case

**Comment by zertrin** [comment]
This also explains why you get `BackendException: Boto requires a bucket name.` in your comment, boto (which is the backend speaking to the S3 storage) is expecting a hostname before the bucket name when using `s3://`

**Comment by zertrin** [comment]
Have a look at this pull request too: https://github.com/zertrin/duplicity-backup.sh/pull/40/files

This is the one that introduced these lines.

I especially feel that the part where `# Strip off the host name, too.` is weird... Why would I remove the host name before passing the destination to `${S3CMD} du -H s3://${TMPDEST}` ??

Sadly I cannot test this use case as I don't have any S3-compatible service that is not AWS.

Hope these clues might help you in diagnosing the issue.

**Comment by colin-nolan** [comment]
Thanks very much for your help @zertrin.

My problems were caused by a series of misunderstanding, exacerbated by my neglect to read your comments in the example configuration file.

I incorrectly assumed s3 and s3+http were the same, which, as your comments indicate, is wrong. I also failed to understand how Duplicity would know about my host if I just gave the bucket name (I wishfully thought it was using `s3cmd` behind the scenes and would use the configuration I gave duplicity-backup.sh for that).

The `# Strip off the host name, too.` bit makes sense if `host_bucket` is defined in the `s3cmd` configuration file. If it's not stripped and `host_bucket` is set, `s3cmd` will incorrectly use the host name as the bucket name.

I've worked around my problems by changing to use the s3+http scheme. I think [the pull request that I made](https://github.com/zertrin/duplicity-backup.sh/pull/156) is still worth while as it leads to a much more understandable failure!

As for testing with something that looks like S3 but is not from Amazon, take a look at https://github.com/minio/minio.
